### PR TITLE
Fix error of unique index key order

### PR DESCRIPTION
**Pull request by yaowangm**
Co-authored-by: Xing Guo <admin@higuoxing.com>

### Issue

The unique index key order will affect the distribution key order of child partitions. For example we create a partitioned table with distribution key(a, b, c). After that, an unique index created on (a, c, b):
```
CREATE TABLE foo1 (a int, b int, c int)  DISTRIBUTED BY (a,b,c) PARTITION BY RANGE(a) 
(PARTITION p1 START (1) END (10000) INCLUSIVE,
PARTITION p2 START (10001) END (100000) INCLUSIVE,
PARTITION p3 START (100001) END (1000000) INCLUSIVE);
create unique index acb_idx on public.foo1 using btree(a,c,b);
```
When adding a new partition p4, we found the p4 distribution key has been changed to (a ,c ,b) which is different from other partitions:
```
alter table public.foo1 add partition p4 START (1000001) END (2000000) INCLUSIVE;
\d+ foo1_1_prt_p4
                   Table "public.foo1_1_prt_p4"
 Column |  Type   | Modifiers | Storage | Stats target | Description 
--------+---------+-----------+---------+--------------+-------------
 a      | integer |           | plain   |              | 
 b      | integer |           | plain   |              | 
 c      | integer |           | plain   |              | 
Indexes:
    "foo1_1_prt_p4_a_c_b_idx" UNIQUE, btree (a, c, b)
Check constraints:
    "foo1_1_prt_p4_check" CHECK (a >= 1000001 AND a <= 2000000)
Inherits: foo1
Distributed by: (a, c, b)
```
This causes an inconsistency issue in gpcheckcat:
```
Performing test 'part_integrity'
[ERROR]: child partition(s) are distributed differently from the root partition, and must be manually redistributed, for some tables. Check the gpcheckcat log for details.
```
### Root cause
The issue was imported by https://github.com/greenplum-db/gpdb/pull/10969. The design is to use the most common columns in all unique indexes as the distribution keys, and the set of the distribution keys must be a subset of the set of columns on the unique constraint. See the code in `transformDistributedBy()`:
```
			foreach(cell, index_stmt->indexParams)
			{
				IndexElem *iparam = lfirst(cell);
				ListCell *dkcell;

				/*
				 * The index element could be either a column name or an expression.
				 * If the index element is not a column name, it should be skipped
				 * to compute the most common columns. For example,
				 *
				 *   create table t(i int, j int, k int) distributed by (i,j);
				 *   create unique index on t(i, func1(j));
				 *
				 * The first index element is a name, the second index element
				 * is an expression. The set of distribution keys is not a subset
				 * of the column names in the index, so it violates the
				 * compatibility and finally it fails.
				 * But `create unique index on t(i, j);` will success.
				 */
				if (!iparam || !iparam->name)
					continue;
				foreach(dkcell, distrkeys)
				{
					IndexElem  *dk = (IndexElem *) lfirst(dkcell);
					if (strcmp(dk->name, iparam->name) == 0)
					{
						new_distrkeys = lappend(new_distrkeys, dk);
						break;
					}
				}
			}
```
However, it didn't consider the order of new dist keys, and the order of unique indexes is always used for new dist keys.

### Solution
We can go through dist keys in outer loop and unique indexes in inner loop. Since the target is to get the common subset of them, it can guarantee to get the correct subset with order of dist keys.

### CODE REVIEW

**Comment by yaowangm** [comment]
Some more analysis:

On GPDB6:

1. As I mentioned, the problematic code in `transformDistributedBy()` was imported by https://github.com/greenplum-db/gpdb/pull/10969. According to code comments, the code is to use the most common columns in all unique indexes as the distribution keys when no DISTRIBUTED BY was given. So the order of distribution keys is not important.
2. For current case, `transformDistributedBy()` is called to create new partition for a partition table. Since the new partition must has exact same distribution keys to root partition, the info of distribution keys is given actually. Theoretically, the calculation of new distribution keys in unnecessary. 
3. With current fix,  `transformDistributedBy()` always generate a sub set of distribution keys and unique index/primary key columns. Since distribution keys must be a subset of the set of columns on the unique constraint, it always return existing distribution keys itself (with correct order). So it works well.
4. An (more reasonable) alternative is, we can simply check whether it is to create adding a part to a partition (`is_add_part`), and return the existing distribution keys. **Please see the latest two commits as the alternative.** More tests are under working.

On GPDB7:

1. The case works even without my fix, because `transformDistributedBy()`  is not called at the time. See the code in `transformCreateStmt()`:
```
 404     if (stmt->relKind == RELKIND_RELATION)
 405     {
 406         stmt->distributedBy = transformDistributedBy(pstate, &cxt,
 407                                                      stmt->distributedBy,
 408                                                      likeDistributedBy, bQuiet);
 409     }
```
and `stmt->relKind` was unsigned:
```
(gdb) p stmt->relKind 
$9 = 0 '\000'
```
2. For the case (alter table add partition), GPDB7 has different call stack from 6X. See stack on GPDB7:
```
(gdb) bt
#0  transformCreateStmt (stmt=stmt@entry=0x55d3f1068590, queryString=queryString@entry=0x55d3e9ee7360 "(internally generated SQL command)") at parse_utilcmd.c:188
#1  0x000055d3e96a741d in ProcessUtilitySlow (pstate=pstate@entry=0x55d3f1067e90, pstmt=pstmt@entry=0x55d3f1067fa8, 
    queryString=queryString@entry=0x55d3e9ee7360 "(internally generated SQL command)", context=context@entry=PROCESS_UTILITY_SUBCOMMAND, params=params@entry=0x0, 
    queryEnv=queryEnv@entry=0x0, dest=0x55d3ea30ee40 <donothingDR>, completionTag=0x0) at utility.c:1206
#2  0x000055d3e96a70ab in standard_ProcessUtility (pstmt=pstmt@entry=0x55d3f1067fa8, 
    queryString=queryString@entry=0x55d3e9ee7360 "(internally generated SQL command)", context=context@entry=PROCESS_UTILITY_SUBCOMMAND, params=params@entry=0x0, 
    queryEnv=queryEnv@entry=0x0, dest=dest@entry=0x55d3ea30ee40 <donothingDR>, completionTag=0x0) at utility.c:1121
#3  0x000055d3e96a726e in ProcessUtility (pstmt=0x55d3f1067fa8, queryString=0x55d3e9ee7360 "(internally generated SQL command)", 
    context=context@entry=PROCESS_UTILITY_SUBCOMMAND, params=params@entry=0x0, queryEnv=queryEnv@entry=0x0, dest=0x55d3ea30ee40 <donothingDR>, completionTag=0x0)
    at utility.c:401
#4  0x000055d3e947d58b in ATExecGPPartCmds (origrel=origrel@entry=0x7f3c16895438, cmd=cmd@entry=0x55d3f11707a8) at tablecmds_gp.c:1475
#5  0x000055d3e944cdf2 in ATExecCmd (wqueue=wqueue@entry=0x7ffc76ac1e10, tab=tab@entry=0x55d3f105c9f0, rel=rel@entry=0x7f3c16895438, cmd=0x55d3f11707a8, 
    lockmode=lockmode@entry=8) at tablecmds.c:5782
#6  0x000055d3e944ce2e in ATRewriteCatalogs (wqueue=wqueue@entry=0x7ffc76ac1e10, lockmode=lockmode@entry=8) at ../../../src/include/nodes/nodes.h:715
#7  0x000055d3e944d02f in ATController (parsetree=parsetree@entry=0x55d3f105a188, rel=rel@entry=0x7f3c16895438, cmds=<optimized out>, recurse=<optimized out>, 
    lockmode=lockmode@entry=8) at tablecmds.c:4846
#8  0x000055d3e944d0b8 in AlterTable (relid=relid@entry=17162, lockmode=lockmode@entry=8, stmt=stmt@entry=0x55d3f105a188) at tablecmds.c:4276
#9  0x000055d3e96a78a6 in ProcessUtilitySlow (pstate=pstate@entry=0x55d3f105a070, pstmt=pstmt@entry=0x55d3f103a128, 
    queryString=queryString@entry=0x55d3f1039060 "alter table public.foo1 add partition p4 START (1000001) END (2000000) INCLUSIVE;", 
    context=context@entry=PROCESS_UTILITY_TOPLEVEL, params=params@entry=0x0, queryEnv=queryEnv@entry=0x0, dest=0x55d3f103a418, completionTag=0x7ffc76ac2520 "")
    at utility.c:1467
#10 0x000055d3e96a70ab in standard_ProcessUtility (pstmt=pstmt@entry=0x55d3f103a128, 
    queryString=queryString@entry=0x55d3f1039060 "alter table public.foo1 add partition p4 START (1000001) END (2000000) INCLUSIVE;", 
    context=context@entry=PROCESS_UTILITY_TOPLEVEL, params=params@entry=0x0, queryEnv=queryEnv@entry=0x0, dest=dest@entry=0x55d3f103a418, 
    completionTag=0x7ffc76ac2520 "") at utility.c:1121
...
```
And GPDB6:
```
#0  transformDistributedBy (cxt=cxt@entry=0x7fff05a328d0, distributedBy=<optimized out>, likeDistributedBy=likeDistributedBy@entry=0x0, bQuiet=0 '\000')
    at parse_utilcmd.c:2135
#1  0x000055f8943341f0 in transformCreateStmt (stmt=0x55f89c3e99e8, stmt@entry=0x55f89c5487e0, queryString=queryString@entry=0x55f894f2ad4f "ADD PARTITION", 
    createPartition=createPartition@entry=1 '\001') at parse_utilcmd.c:459
#2  0x000055f8947b0da0 in atpxPartAddList (rel=rel@entry=0x7f90a9917d48, is_split=is_split@entry=0 '\000', colencs=colencs@entry=0x0, pNode=0x55f89c3e9b60, 
    partName=0x55f89c548a50 "p4", isDefault=<optimized out>, pelem=0x55f89c4d2f30, part_type=PARTTYP_RANGE, par_prule=0x0, 
    lrelname=0x7fff05a32ba0 "relation \"foo1\"", bSetTemplate=0 '\000', ownerid=10) at cdbpartition.c:7025
#3  0x000055f89439c33a in ATPExecPartAdd (tab=tab@entry=0x55f89c5486c8, rel=rel@entry=0x7f90a9917d48, pc=<optimized out>, att=AT_PartAdd) at tablecmds.c:16248
#4  0x000055f8943b5294 in ATExecCmd (wqueue=wqueue@entry=0x7fff05a32d10, tab=tab@entry=0x55f89c5486c8, rel_p=rel_p@entry=0x7fff05a32cb0, 
    cmd=cmd@entry=0x55f89c4d2e98, lockmode=lockmode@entry=8) at tablecmds.c:5744
#5  0x000055f8943b53d8 in ATRewriteCatalogs (wqueue=wqueue@entry=0x7fff05a32d10, lockmode=lockmode@entry=8) at tablecmds.c:5466
#6  0x000055f8943b5629 in ATController (rel=rel@entry=0x7f90a9917d48, cmds=<optimized out>, recurse=1 '\001', lockmode=lockmode@entry=8) at tablecmds.c:4147
#7  0x000055f8943b5788 in AlterTable (relid=relid@entry=16505, lockmode=lockmode@entry=8, stmt=0x55f89c4d3178) at tablecmds.c:3779
#8  0x000055f8945bcf26 in ProcessUtilitySlow (parsetree=parsetree@entry=0x55f89c3e47e8, 
    queryString=queryString@entry=0x55f89c3e3798 "alter table public.foo1 add partition p4 START (1000001) END (2000000) INCLUSIVE;", 
    context=context@entry=PROCESS_UTILITY_TOPLEVEL, params=params@entry=0x0, dest=dest@entry=0x55f89c3e4b70, completionTag=completionTag@entry=0x7fff05a331c0 "")
    at utility.c:1333
#9  0x000055f8945bc96a in standard_ProcessUtility (parsetree=0x55f89c3e47e8, 
    queryString=0x55f89c3e3798 "alter table public.foo1 add partition p4 START (1000001) END (2000000) INCLUSIVE;", context=PROCESS_UTILITY_TOPLEVEL, params=0x0, 
    dest=0x55f89c3e4b70, completionTag=0x7fff05a331c0 "") at utility.c:1049
```
3. Instead of calling `atpxPartAddList()` by query string "ADD PARTITION" , GPDB7 uses kind of "(internally generated SQL command)", and `stmt->relKind` is unavailable for the case. It makes sense and avoids the problematic code in `transformDistributedBy()`. So **no fix is needed for GPDB7**.

@higuoxing Please give your comments, thanks!

File: src/backend/parser/parse_utilcmd.c [hunk 1]
```diff
@@ -423,8 +423,11 @@ transformCreateStmt(CreateStmt *stmt, const char *queryString, bool createPartit
 	/*
 	 * Transform DISTRIBUTED BY (or constuct a default one, if not given
 	 *  explicitly). Not for foreign tables, though.
+	 * If it is to add a part to a partition table, the new partition must
+	 * has exact same distribution keys to root partition. So the transformation
+	 * is unnecessary.
 	 */
-	if (stmt->relKind == RELKIND_RELATION)
+	if (stmt->relKind == RELKIND_RELATION && !stmt->is_add_part)
```
Comment (higuoxing) [inline src/backend/parser/parse_utilcmd.c:430]: Though the `stmt->is_split_part` is not used in 6X, can we mention it here to indicate that we have considered the 'SPLIT PARTITION' situation?

**Review by higuoxing — APPROVED** [review]
Looks good to me!

File: src/backend/parser/parse_utilcmd.c (hunk 1 above)
Comment (yaowangm) [reply src/backend/parser/parse_utilcmd.c:433]: Good idea! I have updated comments.

**Comment by yaowangm** [comment]
Question by @kainwen: Is there any side effect if we ignore the code block for `stmt->is_add_part` in `transformCreateStmt()`, especially the calling of `transformDistributedBy()`?

Answer:

1. I scanned all code in the ignored block. All of them, except `transformDistributedBy()`, are related to `stmt->is_part_child` or `stmt->is_part_parent`, which is conflict to `stmt->is_add_part`. So we can ignore them safely.
2. `In transformDistributedBy()`, we can safely return `distributedBy` itself in the beginning if no change is needed. There is an existing example (`distributedBy->numsegments` is not changed actually):
```
1860     /* Explictly specified distributed randomly, no futher check needed */
1861     if (distributedBy &&
1862         (distributedBy->ptype == POLICYTYPE_PARTITIONED && distributedBy->keyCols == NIL))
1863     {
1864         distributedBy->numsegments = numsegments;
1865         return distributedBy;
1866     }
```
The reason why I ignore the calling of `transformDistributedBy()` but not change `transformDistributedBy()` itself is that `stmt->is_add_part` is unavailable inside `transformDistributedBy()`.
3. The line to control the ignorance:
```
    if (stmt->relKind == RELKIND_RELATION)
```
was imported in https://github.com/greenplum-db/gpdb/commit/25a90396cd1c52d252a37986e12b63e8e037aa83, which is to merge with PostgreSQL 9.1 beta2 on May 19, 2018, a pretty old date.
4. The case test and CI pipeline works well with the fix.

So I didn't find any side effect. @higuoxing Please supplement if necessary, thx!

**Comment by higuoxing** [comment]
The `is_add_part` flag and the `distributedBy` clause are initialized together. IMHO, if there're side effects, we should fix the function `make_distributedby_for_rel(Relation rel)`. I think this PR is safe to merge.

See:
https://github.com/greenplum-db/gpdb/blob/f12c2f0711e6a28c34958a7477f359c9eec68305/src/backend/cdb/cdbpartition.c#L6806-L6814

**Comment by kainwen** [comment]
Thanks both. Please go ahead and push this. @yaowangm @higuoxing

**Pull request merged**